engine/uState.ml

```diff
@@ -169,10 +170,14 @@ let add ~check_fresh ~named q m =
 let of_set qs =
   { named = QSet.empty; qmap = QMap.bind (fun _ -> None) qs; above = QSet.empty }
 
+let of_named qs = 
+  { (of_set qs) with
+    named = qs }
+
 (* XXX what about [above]? *)
 let undefined m =
-  let m = QMap.filter (fun _ v -> Option.is_empty v) m.qmap in
-  QMap.domain m
+  let mq = QMap.filter (fun _ v -> Option.is_empty v) m.qmap in
+  QSet.diff (QMap.domain mq) m.named
```

Comment:
I'm not sure this is the right approach, probably QState should know which variables are local and named, and which are global. Local named variables shouldn't be removed from undefined.